File: app/src/main/java/at/tomtasche/reader/nonfree/BillingManager.java
package at.tomtasche.reader.nonfree;

import android.content.Context;

import at.tomtasche.reader.BuildConfig;
import at.tomtasche.reader.background.BillingPreferences;

public class BillingManager {

    private boolean enabled;

    private AdManager adManager;

    private BillingPreferences billingPreferences;

    public void initialize(Context context, AnalyticsManager analyticsManager, AdManager adManager) {
        this.adManager = adManager;

        billingPreferences = new BillingPreferences(context);

        if (BuildConfig.FLAVOR.equals("pro")) {
            billingPreferences.setPurchased(true);
        }

        if (!enabled) {
            adManager.showGoogleAds();
            return;
        }

        enforceAds();
    }

    private void enforceAds() {
        if (hasPurchased()) {
            adManager.removeAds();
        } else {
            adManager.showGoogleAds();
        }
    }

    public boolean hasPurchased() {
        if (!enabled) {
            return true;
        }

        if (billingPreferences == null) {
            return false;
        }

        return billingPreferences.hasPurchased();
    }

    public void setEnabled(boolean enabled) {
        this.enabled = enabled;
    }

    public boolean isEnabled() {
        return enabled;
    }
}

File: app/src/main/java/at/tomtasche/reader/nonfree/BillingManager.kt
package at.tomtasche.reader.nonfree

import android.content.Context
import at.tomtasche.reader.BuildConfig
import at.tomtasche.reader.background.BillingPreferences

class BillingManager {

    private var enabled = false

    private var adManager: AdManager? = null
    private var billingPreferences: BillingPreferences? = null

    fun initialize(context: Context, analyticsManager: AnalyticsManager, adManager: AdManager) {
        this.adManager = adManager

        val preferences = BillingPreferences(context)
        billingPreferences = preferences

        if (BuildConfig.FLAVOR == "pro") {
            preferences.setPurchased(true)
        }

        if (!enabled) {
            adManager.showGoogleAds()
            return
        }

        if (hasPurchased()) {
            adManager.removeAds()
        } else {
            adManager.showGoogleAds()
        }
    }

    fun hasPurchased(): Boolean {
        if (!enabled) {
            return true
        }

        return billingPreferences?.hasPurchased() ?: false
    }

    fun setEnabled(enabled: Boolean) {
        this.enabled = enabled
    }

    fun isEnabled(): Boolean = enabled
}
